pkg/reconciler/managed/reconciler.go

```diff
@@ -1248,3 +1291,92 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (resu
 	managed.SetConditions(xpv1.ReconcileSuccess())
 	return reconcile.Result{RequeueAfter: reconcileAfter}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 }
+
+// recordChangeLog constructs a change log entry from the provided information
+// and sends it to the change log service, if the feature is enabled. If errors
+// are encountered during this process, then logs/events will be recorded for
+// each, but the process will continue. We are choosing to send all the
+// information we can to the change logs as opposed to sending nothing.
+func (r *Reconciler) recordChangeLog(ctx context.Context, managed resource.Managed, snapshot *structpb.Struct, log logging.Logger,
+	record event.Recorder, opType protov1alpha1.OperationType, changeErr error, ad *AdditionalDetails,
+) {
+	if !r.features.Enabled(feature.EnableAlphaChangeLogs) {
+		// change log feature isn't enabled, just return immediately
+		return
+	}
+
+	// convert any additional details to a Struct to include in the change log entry
+	var ads *structpb.Struct
+	if ad != nil {
+		b, err := json.Marshal(ad)
+		if err != nil {
+			log.Debug("Cannot marshal additional details", "error", err)
+			record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
+		} else {
+			ads = &structpb.Struct{}
+			if err = ads.UnmarshalJSON(b); err != nil {
+				log.Debug("Cannot convert additional details to Struct", "error", err)
+				record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
+				ads = nil
+			}
+		}
+	}
+
+	// determine the full namespaced name of the managed resource
+	var namespacedName string
```

Comment:
`namespacedName := types.NamespacedName{Namespace: managed.GetNamespace(), Name: managed.GetName()}.String()`

Reply:
yep yep, I actually tried that first too, but didn't like the output for when there is no namespace. [`NamespacedName.String()`](https://github.com/kubernetes/apimachinery/blob/master/pkg/types/namespacedname.go#L38 ) always uses a separator char even when there is no namespace, so the output looks like `/my-resource`.

I was doubting even including the namespace at all because managed resources aren't namespaced right now in Crossplane v1.  Perhaps that will change for a v2, so I may have been getting ahead of myself with future-proofing by including handling namespaces already at this point 😅 

I could just remove this entirely and include only the name of the resource, without any consideration for namespace.

Reply:
Namespace usage removed in https://github.com/crossplane/crossplane-runtime/pull/754/commits/c23f09b8314e5a7d11e0438243572a0b38cfaff1 as per https://github.com/crossplane/crossplane-runtime/pull/754#discussion_r1707711834.